libraries/ms-drivers/inc/bts_7200_current_sense.h

```diff
@@ -4,49 +4,104 @@
+// only be manipulated through this driver.
+
 #include "adc.h"
 #include "gpio.h"
 #include "pca9539r_gpio_expander.h"
 #include "soft_timer.h"
 
+// Upper maximum for the possible leakage voltage that may be read from the SENSE pin at
+// T(env) < 80 C (see p.g. 27 of BTS7200 datasheet)
+#define BTS7200_MAX_LEAKAGE_VOLTAGE_MV 2
+
+// Nominal scaling factor k(ILIS) for current output at the SENSE pin in normal operation.
+#define BTS7200_IS_SCALING_NOMINAL 670
+
+// Voltage at the SENSE pin is limited to a max of 3.3V by a diode.
+// Due to to this function, since any fault current will be at least 4.4 mA (see p.g. 49)
+// the resulting voltage will be 4.4 mA * 1.6 kOhm = ~7 V. Due to this,
+// voltages approaching 3.3V represent a fault, and should be treated as such.
+#define BTS7200_MAX_VALID_SENSE_VOLTAGE_MV 3200
+
+// Check if measurement is a fault.
+#define BTS7200_IS_MEASUREMENT_FAULT_MV(meas) ((meas) > BTS7200_MAX_VALID_SENSE_VOLTAGE_MV)
+
+// Max possible delay after input pin pulled low on fault, + 10 ms for buffer
+#define BTS7200_FAULT_RESTART_DELAY_MS 110
+#define BTS7200_FAULT_RESTART_DELAY_US (BTS7200_FAULT_RESTART_DELAY_MS * 1000)
+
 typedef void (*Bts7200DataCallback)(uint16_t reading_out_0, uint16_t reading_out_1, void *context);
 
+typedef void (*Bts7200FaultCallback)(bool fault0, bool fault1, void *context);
+
+// Represents whether SEL/EN pins are accessed through STM32 or a Pca9539R
+typedef enum {
+  BTS7200_PIN_STM32 = 0,
+  BTS7200_PIN_PCA9539R,
+  NUM_BTS7200_PIN_TYPES,
+} Bts7200PinType;
+
+// Holds pin-specific info for EN pins
+typedef struct {
+  GpioAddress *enable_pin_stm32;
+  Pca9539rGpioAddress *enable_pin_pca9539r;
+  SoftTimerId fault_timer_id;
+  bool fault_in_progress;
+  Bts7200PinType pin_type;
+} Bts7200EnablePin;
+
+// Holds pin-specific info for SEL pins
+typedef struct {
+  GpioAddress *select_pin_stm32;
+  Pca9539rGpioAddress *select_pin_pca9539r;
+  Bts7200PinType pin_type;
+} Bts7200SelectPin;
+
 // Use when the select pin is an STM32 GPIO pin
 typedef struct {
   GpioAddress *select_pin;
   GpioAddress *sense_pin;
+  GpioAddress *enable_0_pin;
+  GpioAddress *enable_1_pin;
   uint32_t interval_us;
   Bts7200DataCallback callback;  // set to NULL for no callback
   void *callback_context;
+  Bts7200FaultCallback fault_callback;
+  void *fault_callback_context;
+  int resistor;  // scaling performed by resistor at SENSE pin
 } Bts7200Stm32Settings;
 
 // Use when the select pin is through a PCA9539R GPIO expander
 typedef struct {
   I2CPort i2c_port;
   Pca9539rGpioAddress *select_pin;
   GpioAddress *sense_pin;
+  Pca9539rGpioAddress *enable_0_pin;
+  Pca9539rGpioAddress *enable_1_pin;
   uint32_t interval_us;
   Bts7200DataCallback callback;  // set to NULL for no callback
   void *callback_context;
+  Bts7200FaultCallback fault_callback;
+  void *fault_callback_context;
+  int resistor;  // scaling performed by resistor at SENSE pin
```

Comment:
Ah, I know I said int is okay but could you actually use a `uint32_t` or something like that (and in all the other places you set resistor values)? Sorry for the confusion.

You should also specify that it's in milliohms.